##  A genome-wide gene expression analysis in APP-transgenic mice during development

***

```{r, echo=F, warning=F, message=F}
library(ape)
library(amap)
library(ggdendro)
library(MASS)
library(gplots)
library(ggplot2)
library(ggvis)
library(pheatmap)
library(grid)
library(dplyr)
library(contrast)
library(multcomp)
library(png)

rm(list = ls())
setwd("~/Dropbox/GitHub/Ad")

load("./data/brain2014.rdt"); glm.dt <- brain.tpm
# load("./data/retina2014.rdt"); glm.dt <- retina.tpm
source("function.R")
```

### We used 2014 samples only, and kept genes that fit two criterias:

1. Maximal expression needs be bigger than 0.25 quantile of the complete datasets
2. Expression value needs be non-zero in at least 1/10 of the samples.

```{r}
cutoff <- quantile(c(as.matrix(glm.dt)), 0.25)  # TPM level
glm.dt <- glm.dt[apply(glm.dt, 1, function(x) max(x) > cutoff & sum(x > 0) > round(ncol(glm.dt) / 10)), ]

dim(glm.dt)
par(mar = c(5, 5, 0, 2))
hist(c(as.matrix(glm.dt)), main = NULL, xlab = "Log2(TPM)", col = "dodgerblue3", border = "white")
```

### Model the data with two GLMs

We took off each gene's batch effect as estimated by a GLM 4 predictors: age, group, batch, and age group interaction.

$y \sim age + group + batch + age * group$

```{r}
age <- factor(gsub("^.*(2m|4m|5m|6m).*", "\\1", colnames(glm.dt)), levels = c("2m", "4m", "5m", "6m"))
group <- factor(gsub("^.*(WT|APP).*", "\\1", colnames(glm.dt)), levels = c("WT", "APP"))
uid <- paste(age, group, sep = "_"); table(uid)
batch <- factor(gsub("^.*(2014|mouse).*", "\\1", colnames(glm.dt)), levels = c("2014", "mouse"))
conditions <- c("2m_WT", "2m_APP", "4m_WT", "4m_APP", "5m_WT", "5m_APP", "6m_WT", "6m_APP")
conditions <- conditions[conditions %in% uid]

fit0 <- apply(glm.dt, 1, function (x) summary(lm(x ~ age + group + batch + age*group)))
fit0.batch <- sapply(fit0, function (x) x$coefficients["batchmouse", "Estimate"])
glm.dt_bc <- glm.dt - as.matrix(fit0.batch) %*% (as.numeric(batch) - 1)
```

The batch-corrected gene expressions were modeled by anotehr GLM with 3 predictors: age, group, and age group interaction.

$y \sim age + group + age * group$

```{r}
glm.fit <- apply(glm.dt_bc, 1, function (x) lm(x ~ age + group + age*group))
summary(glm.fit[["App"]])$coefficient
```

### Choose genes with R2 > 0.3 and Qvalue < 0.05

GLM's R2 and F-test q-value define how well each gene's expression could be predicted by the predictors. Small p-value and big R2 suggest good fit.

```{r}
fit.r2 <- sapply(glm.fit, function (x) summary(x)$r.squared)
fit.qval <- sapply(glm.fit, function (x) summary(x)$fstatistic) %>% 
  apply(2, function (x) pf(x[1], x[2], x[3], lower.tail = F)) %>% p.adjust(method = "fdr")

glm.fit <- glm.fit[fit.qval < 0.05 & fit.r2 > 0.3]
length(glm.fit)
```

### Age, App, and age_App_inter genes with ANOVA

ANOVA identifies the normal aging genes, 2-month app genes, age-dependent app genes (4, 5, 6 months)

```{r}
fit.anova <- lapply(glm.fit, anova)
fit.anova[["App"]]

genes.aov <- lapply(c("age", "group", "age:group"), function(x1) {
  y1 = sapply(fit.anova, function(x2) x2[x1, "Pr(>F)"]); y1[y1 < 0.05]
}); names(genes.aov) <- c("age", "app", "age:app")

sapply(genes.aov, function(x) length(x))
par(mar = c(0, 5, 0, 2))
venn(list(age = names(genes.aov[[1]]), APP = names(genes.aov[[2]]), age_int_app = names(genes.aov[[3]])))
```

### Age-specific APP effect

This is equivalent with pairwise t-tests between WT and APP samples for each month.

```{r}
age_spe <- lapply(c("2m", "4m", "5m", "6m"), function(x1) {
  sapply(glm.fit, function(x2) 
    contrast(x2, list(age = x1, group = "APP"), list(age = x1, group = "WT")))[c("Contrast", "Pvalue"), ]
}); names(age_spe) <- c("2m", "4m", "5m", "6m")

sapply(c("2m", "4m", "5m", "6m"), function(x1) {
  contrast(glm.fit[["Stat3"]], list(age = x1, group = "APP"), list (age = x1, group = "WT"))  
})[c("Contrast", "Pvalue"), ]

age_spe_up <- sapply(age_spe, function(x) x[, x[1, ] > 0.1 & x[2, ] < 0.05])
age_spe_dn <- sapply(age_spe, function(x) x[, x[1, ] < -0.1 & x[2, ] < 0.05])

sapply(age_spe_up, function(x) ncol(x)) 
sapply(age_spe_dn, function(x) ncol(x))
```

### WT-specific age effect (normal aging genes)

```{r}
combn <- combn(c("2m", "4m", "5m", "6m"), 2)
geno_spe_wt <- lapply(1:ncol(combn), function(x1) {
  sapply(glm.fit, function(x2)
    contrast(x2, list(age = combn[2, x1], group = "WT"), list(age = combn[1, x1], group = "WT")))[c("Contrast", "Pvalue"), ]
}); names(geno_spe_wt) <- apply(combn, 2, function(x) paste0(x[2], "_vs_", x[1])) 

sapply(1:ncol(combn), function(x1) {
  contrast(glm.fit[["Stat3"]], list(age = combn[2, x1], group = "WT"), list(age = combn[1, x1], group = "WT")) 
})[c("Contrast", "Pvalue"), ]

geno_spe_wt_up <- sapply(geno_spe_wt, function(x) x[, x[1, ] > 0.1 & x[2, ] < 0.05])
geno_spe_wt_dn <- sapply(geno_spe_wt, function(x) x[, x[1, ] < -0.1 & x[2, ] < 0.05])

sapply(geno_spe_wt_up, function(x) ncol(x)) 
sapply(geno_spe_wt_dn, function(x) ncol(x))
```

### APP-specific age effect (app-specific aging genes)

```{r}
C <- matrix(c(0,0,0,0,0,0,-1,1), 1) 
cat("6m APP vs 5m APP Stat3: \n"); summary(glht(glm.fit[["Stat3"]], linfct = C))
```

<!--

glmList <- list()
glmList$age_spe_up <- age_spe_up
glmList$age_spe_dn <- age_spe_dn
glmList$geno_spe_wt_up <- geno_spe_wt_up
glmList$geno_spe_wt_dn <- geno_spe_wt_dn

save(glmList, file = "data/glmList.rdt")

-->
